support-frontend/assets/components/subscriptionCheckouts/addressSearch/addressComponentStore.js

```diff
@@ -0,0 +1,262 @@
+import { postcodeIsWithinDeliveryArea } from 'helpers/deliveryCheck';
+import type { FulfilmentOptions } from 'helpers/productPrice/fulfilmentOptions';
+
+// ----- Types ----- //
+
+export type FormFields = {|
+  ...RegularPaymentRequestAddress
+|};
+export type FormField = $Keys<FormFields>;
+export type FormErrors = FormError<FormField>[];
+
+type AddressFieldsState = {|
+  ...FormFields,
+  formErrors: FormErrors,
+|}
+
+export type State = {|
+  fields: AddressFieldsState,
+  postcode: PostcodeFinderState
+|};
+
+export type SetCountryChangedAction = { type: 'SET_COUNTRY_CHANGED', country: IsoCountry, ...Scoped<AddressType> };
+
+export type Action =
+  | { type: 'SET_ADDRESS_LINE_1', lineOne: string, ...Scoped<AddressType> }
+  | { type: 'SET_ADDRESS_LINE_2', lineTwo: string, ...Scoped<AddressType> }
+  | { type: 'SET_TOWN_CITY', city: string, ...Scoped<AddressType> }
+  | { type: 'SET_STATE', state: string, ...Scoped<AddressType> }
+  | SetCountryChangedAction
+  | { type: 'SET_ADDRESS_FORM_ERRORS', errors: FormError<FormField>[], ...Scoped<AddressType> }
+  | { type: 'SET_POSTCODE', postCode: string, ...Scoped<AddressType> };
+
+
+// ----- Selectors ----- //
+
+const getPostcodeForm = (state: State): PostcodeFinderState => state.postcode;
+const getStateFormErrors = (state: State): FormErrors => state.fields.formErrors;
+
+const getFormFields = (state: State): FormFields => ({
+  lineOne: state.fields.lineOne,
+  lineTwo: state.fields.lineTwo,
+  city: state.fields.city,
+  country: state.fields.country,
+  postCode: state.fields.postCode,
+  state: state.fields.state,
+});
+
+
+// ----- Functions ----- //
+
+const isPostcodeOptional = (country: Option<IsoCountry>): boolean =>
+  country !== 'GB' && country !== 'AU' && country !== 'US' && country !== 'CA';
+
+const checkpostCodeLength = (input: string | null): boolean => ((input == null) || (input.length <= 20));
+
+const isStateNullable = (country: Option<IsoCountry>): boolean =>
+  country !== 'AU' && country !== 'US' && country !== 'CA';
+
+export const isHomeDeliveryInM25 = (fulfilmentOption: Option<FulfilmentOptions>, postcode: Option<string>) => {
+  if (fulfilmentOption === 'HomeDelivery' && postcode !== null) {
+    return postcodeIsWithinDeliveryArea(postcode);
+  }
+  return true;
+};
+
+const applyBillingAddressRules = (fields: FormFields, addressType: AddressType): FormError<FormField>[] => validate([
+  {
+    rule: nonEmptyString(fields.lineOne),
+    error: formError('lineOne', `Please enter a ${addressType} address.`),
+  },
+  {
+    rule: nonEmptyString(fields.city),
+    error: formError('city', `Please enter a ${addressType} city.`),
+  },
+  {
+    rule: isPostcodeOptional(fields.country) || nonEmptyString(fields.postCode),
+    error: formError('postCode', `Please enter a ${addressType} postcode.`),
+  },
+  {
+    rule: checkpostCodeLength(fields.postCode),
+    error: formError('postCode', `Please enter a ${addressType} postcode no longer than 20 characters.`),
```

Comment:
is this limit for input into the postcode lookup only?  I only ask because there is a card for applying SF limits on city length into the UI, as it causes step function failures. If it's something that goes here we can do it at the same time. https://trello.com/c/SvZwZrTB/2936-billing-city-delivery-instructions-length-restrictions

Reply:
Have done this in another PR

Reply:
fab work that's been hanging around far too long 🏅